### PR TITLE
Dell: Add document.

### PR DESCRIPTION
**Pull request by wang-x-xia**
The PR has 3 parts:
1. Add a document of ECS how-to.
2. Fix the bugs when I test the catalog in production.
3. Add the `:iceberg:dell` module to the runtime package.

### CODE REVIEW

**Comment by rdblue** [comment]
@samredai, can you take a look?

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:29]: Would you mind to provide more description or dell doc link for those two features ? I think we need to let others to know more about the difference between Dell ECS and other vendor Object Storage Service.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:69]: Nit:  `test` -> `<Your-ecs-s3-access-key>`.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:70]: Nit: `xxxxxxxxxxxxxxxx` -> `Your-ecs-s3-secret-access-key`.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:25]: Please add a sentence to say that the `Dell ECS Integration` is available since apache iceberg 0.14.0.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:100]: Since we still don't release the iceberg 0.14.0, so we can not verify & experience this Dell ECS integration work.  It's possible to `wget` the latest snapshot artifact from the maven repo ?

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:119]: Let's put the maven urls and file name into a map so that we don't need to attach the const concated strings in the download code piece ?

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:116]: Will we put this artifact in `//iceberg-flink-runtime-1.14-0.14.0.jar` ? Is it correct ?

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:142]: Is possible to provide a section for the HiveCatalog + Dell ECS ?  In my real experience,  there are many people whose just want to replace the HDFS storage with Object Storage Service while still maintain their tables in HiveCatalog.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:146]: Is the DELL ECS catalog the expected catalog that you Dell ECS guys recommend to ?  In real cases,  I see people will prefer to use HiveCatalog or other centralized metastore to collect all the table so that they can track table transforms in the global view.

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:27]: There may be a better word than session to use here. I see that you mean a sort of reading session but I think that word is used in a lot of other places to mean specific things (like a Spark session). Maybe just say "Iceberg can be used with Dell's Enterprise Object Storage (ECS) by using the ECS catalog."

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:29]: The line above mentions "several features" but then only lists 2. Does it make more sense to provide a more comprehensive list of features or to reword the above sentence to say something like:

> Dell ECS has many features that make Iceberg a highly compatible table format, such as append operations for file writers and content addressable storage (CAS) for table commits.

...then you can exclude this numbered list. What do you think?

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:51]: nit: Maybe add backticks around the `/`?

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:41]: nit: Replace "Username" with "ECS Username"

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:44]: This looks like a typo or grammar issue. Maybe change this to:

> An ECS catalog requires that you configure a warehouse location where all data and metadata will be created.

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:25]: This h2 title and the h1 title above it feels redundant. I would consolidate them into a single title such as `# Dell ECS` or `# Dell ECS Catalog`. Also I think the word "Integration" refers more to Iceberg's incorporation into compute/query engines such as Spark or Trino. I could be wrong but I think Dell ECS is more of an "Implementation".

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:36]: I could be overthinking this but `try` doesn't feel confident enough. I would just say "When using Dell ECS with Iceberg, these configuration parameters are required."

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:53]: "Runtime Dependencies" seems like the more common way this is described.

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:97]: I don't think "For example,` is needed here, I would just jump right in and say "Using the Dell ECS catalog with Flink, you first must create a Flink environment."

File: docs/integrations/dell.md
Comment (samredai) [inline docs/integrations/dell.md:142]: Then, **you can run** `USE CATALOG my_catalog`, `SHOW DATABASES`, and `SHOW TABLES` to fetch the namespaces and tables in the catalog.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:25]: In previous PRs, the reviewers let me use `dell` as the module name.
I also used `dell` as the document title and used the h2 title for ECS. Should I change this document to ECS only?

And the "Integration" will change to "Implementation".

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:146]: Yes. We wish to provide an all-in-one product.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:100]: Because the dependencies are also added in this commit. I don't know which version should I use?

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: The `work_space=/` is an example. I replaced it.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: I've removed the example of `spark-sql` and Flink `sql-client.sh` to avoid redundancy.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: I've changed it.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Done.

**Comment by wang-x-xia** [comment]
Thanks for @samredai 's help. My English is bad. If you have other suggestions, I'll fix them to make this document more readable.

File: docs/integrations/dell.md
Comment (samredai) [reply docs/integrations/dell.md]: On second thought, I think integration makes sense here. I think I'm getting caught up in the fact that this is an implementation of the catalog interface, but you're right--integration is more consistent with how we've been using this elsewhere and overall makes sense (this integrates Iceberg and ECS).

**Review by samredai — APPROVED** [review]
LGTM. I'm thinking more broadly about how we organize integrations into various tools, frameworks, and vendors to make it more intuitive for readers. I may end up including some refactoring/relocating here but that shouldn't hold up this PR.

File: dell/src/main/java/org/apache/iceberg/dell/DellClientFactory.java
Comment (rdblue) [inline dell/src/main/java/org/apache/iceberg/dell/DellClientFactory.java:26]: Can you move code changes out of a documentation PR? These should be in a code PR for people that cherry-pick commits.

File: spark/v3.2/build.gradle
Comment (rdblue) [inline spark/v3.2/build.gradle:216]: Adding this module to the Spark runtime should be discussed and done in a separate PR, not in a documentation PR. Can you bring this up on the dev list and summarize what it pulls in and changes?

You'll also need to update the LICENSE and NOTICE files.

**Review by rdblue — CHANGES_REQUESTED** [review]
I think the documentation is probably close to being ready, but this should not mix code changes, runtime Jar additions, and documentation in a single PR.

**Comment by wang-x-xia** [comment]
@rdblue
The changes are reverted,
I've created a new PR for bugfix:
https://github.com/apache/iceberg/pull/5059
And a new PR for runtime JAR:
https://github.com/apache/iceberg/pull/5060

**Comment by rdblue** [comment]
Since this PR is replaced by newer ones, I'm going to close it.

**Comment by wang-x-xia** [comment]
@rdblue 
This PR is for documents only. I moved other changes to new PRs except the document.
Should I create a new PR for the ducoment?

**Comment by rdblue** [comment]
No, if this is only docs, then we can review it. Thanks, @wang-x-xia.

**Comment by wang-x-xia** [comment]
@rdblue If you have some comments about the document, please let me know. This PR needs your approval now.😂

File: docs/integrations/dell.md
Comment (rdblue) [inline docs/integrations/dell.md:29]: I'm not sure what it means to be a "highly compatible table format". I think this should just have the second sentence, "See Dell ECS for more information."

File: docs/integrations/dell.md
Comment (rdblue) [inline docs/integrations/dell.md:33]: Where?

File: docs/integrations/dell.md
Comment (rdblue) [inline docs/integrations/dell.md:48]: Can you specifically state all of the required properties and where to set them? I think these are catalog properties.

File: docs/integrations/dell.md
Comment (rdblue) [inline docs/integrations/dell.md:52]: This shouldn't start a section. It's probably a good note (if your version is different, make sure you have the right runtime Jar!).

File: docs/integrations/dell.md
Comment (rdblue) [inline docs/integrations/dell.md:64]: Isn't the `iceberg-dell` Jar required as well?

File: docs/integrations/dell.md
Comment (rdblue) [inline docs/integrations/dell.md:131]: I don't think that table rename is supported. Wouldn't that break all metadata pointers?

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:29]: Removed!

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:48]: The following examples will show how to config these properties.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:52]: Removed!

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:64]: Added.
When I wrote this document, I assume all PRs will be merged. 
But now no PR be merged into 0.14.0. So I also fix the version to 0.15.0.
😂

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md:131]: It's just a rename for the catalog layer.
Only the entry object which store the latest version will be renamed, but not changed.
So the data controlled by Iceberg won't be changed. Any pointers won't be broken.

**Comment by wang-x-xia** [comment]
@rdblue 
Could you please confirm whether the next release is 0.14.1 or 0.15?
If the next release is 0.14.1, I'll change the Jar version to 0.14.1.
And if the document has any other problems, I can also fix them! 😂

**Comment by openinx** [comment]
Sorry for the later response, @wang-x-xia. I remember @wang-x-xia ping me in other IM tools, but I missed to track this document. so sorry.  Let me take a look and get this merged if it's ready.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:1]: Is this document path still valid ?  I saw the current docs directory don't have any integration now ?  would you mind to update this PR ?

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:29]: Would you mind to attach the ECS CAS link for more detailed information ?

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:87]: Is it good enough to provide a python script to download the flink jars?  I see the script will need the `pyflink` and `requests`  dependencies and we don't provide any more details about how to run the python script. I'm afraid most of people won't be able to run this python script successfully in their hosts.

I think it's simple enough to wget the flink jars in two line bash scripts and show the following SQL Execution by using flink SQL.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:79]: The link is 404 now.

File: docs/integrations/dell.md
Comment (openinx) [inline docs/integrations/dell.md:60]: Let's just use java/scalar or bash example for keep consistentence as other pages.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: The document is moved and fixed the header~

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: This part is already removed in the latest version.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Now the scripts are bash-script.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: The link is removed.

File: docs/integrations/dell.md
Comment (wang-x-xia) [reply docs/integrations/dell.md]: Now the scripts are bash-script.

File: docs/dell.md
Comment (openinx) [inline docs/dell.md:85]: Only the `pyspark.sql.SparkSession.catalog` has this problems ?  How about the java & scala ?

File: docs/dell.md
Comment (openinx) [inline docs/dell.md:128]: Open question: Should we also add hive section ? I guess many people still use hive tez to process their data, it's okay if you think we don't need to. Thanks.

File: docs/dell.md
Comment (wang-x-xia) [reply docs/dell.md]: All languages are the same. I have fixed this statement.

File: docs/dell.md
Comment (wang-x-xia) [reply docs/dell.md:128]: It's a good idea. Relevant tests are being prepared. And I wish to merge this document firstly.

**Review by openinx — APPROVED** [review]
LGTM, thanks @wang-x-xia for the contribution !

**Pull request merged**
